File: python/src/com/jetbrains/python/packaging/setupPy/CreateSetupPyAction.java
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.jetbrains.python.packaging.setupPy;

import com.intellij.ide.IdeView;
import com.intellij.ide.fileTemplates.FileTemplateManager;
import com.intellij.ide.fileTemplates.actions.AttributesDefaults;
import com.intellij.ide.fileTemplates.actions.CreateFromTemplateAction;
import com.intellij.ide.fileTemplates.ui.CreateFromTemplateDialog;
import com.intellij.ide.util.PropertiesComponent;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.projectRoots.Sdk;
import com.intellij.openapi.roots.ModuleRootManager;
import com.intellij.openapi.roots.ProjectFileIndex;
import com.intellij.openapi.util.Comparing;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.VfsUtilCore;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiManager;
import com.intellij.util.SystemProperties;
import com.jetbrains.python.PyBundle;
import com.jetbrains.python.PyPsiPackageUtil;
import com.jetbrains.python.PythonFileType;
import com.jetbrains.python.packaging.PyPackage;
import com.jetbrains.python.packaging.PyPackageManager;
import com.jetbrains.python.packaging.PyPackageUtil;
import com.jetbrains.python.psi.PyUtil;
import com.jetbrains.python.sdk.PythonSdkUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;


public class CreateSetupPyAction extends CreateFromTemplateAction {
  private static final String AUTHOR_PROPERTY = "python.packaging.author";
  private static final String EMAIL_PROPERTY = "python.packaging.author.email";
  static final String SETUP_SCRIPT_TEMPLATE_NAME = "Setup Script";

  public CreateSetupPyAction() {
    super(
      SETUP_SCRIPT_TEMPLATE_NAME, 
      PythonFileType.INSTANCE.getIcon(), 
      () -> FileTemplateManager.getDefaultInstance().getInternalTemplate(SETUP_SCRIPT_TEMPLATE_NAME)
    );
    getTemplatePresentation().setText(PyBundle.message("python.packaging.create.setup.py"));
  }

  @Override
  public void update(@NotNull AnActionEvent e) {
    final Module module = e.getData(PlatformCoreDataKeys.MODULE);
    e.getPresentation().setEnabled(module != null && !PyPackageUtil.hasSetupPy(module));
  }

  @Override
  public AttributesDefaults getAttributesDefaults(DataContext dataContext) {
    final Project project = CommonDataKeys.PROJECT.getData(dataContext);
    final AttributesDefaults defaults = new AttributesDefaults("setup.py").withFixedName(true);
    if (project != null) {
      defaults.addPredefined("Import", getSetupImport(dataContext));
      defaults.add("Package_name", project.getName());
      final PropertiesComponent properties = PropertiesComponent.getInstance();
      defaults.add("Author", properties.getValue(AUTHOR_PROPERTY, SystemProperties.getUserName()));
      defaults.add("Author_email", properties.getValue(EMAIL_PROPERTY, ""));
      defaults.addPredefined("PackageList", getPackageList(dataContext));
      defaults.addPredefined("PackageDirs", getPackageDirs(dataContext));
    }
    defaults.setAttributeVisibleNames(getVisibleNames());
    return defaults;
  }

  private static Map<String, String> getVisibleNames() {
    HashMap<String, String> attributeToName = new HashMap<>();
    attributeToName.put("Package_name", PyBundle.message("python.packaging.create.setup.package.name"));
    attributeToName.put("Version", PyBundle.message("python.packaging.create.setup.version"));
    attributeToName.put("URL", PyBundle.message("python.packaging.create.setup.url"));
    attributeToName.put("License", PyBundle.message("python.packaging.create.setup.license"));
    attributeToName.put("Author", PyBundle.message("python.packaging.create.setup.author"));
    attributeToName.put("Author_Email", PyBundle.message("python.packaging.create.setup.author.email"));
    attributeToName.put("Description", PyBundle.message("python.packaging.create.setup.description"));
    return attributeToName;
  }

  @NotNull
  private static String getSetupImport(@NotNull DataContext dataContext) {
    final Module module = PlatformCoreDataKeys.MODULE.getData(dataContext);
    return hasSetuptoolsPackage(module) ? "from setuptools import setup" : "from distutils.core import setup";
  }

  private static boolean hasSetuptoolsPackage(@Nullable Module module) {
    final Sdk sdk = PythonSdkUtil.findPythonSdk(module);
    if (sdk == null) return false;

    final List<PyPackage> packages = PyPackageManager.getInstance(sdk).getPackages();
    return packages != null && PyPsiPackageUtil.findPackage(packages, "setuptools") != null;
  }

  private static String getPackageList(DataContext dataContext) {
    final Module module = PlatformCoreDataKeys.MODULE.getData(dataContext);
    if (module != null) {
      return "['" + StringUtil.join(PyPackageUtil.getPackageNames(module), "', '") + "']";
    }
    return "[]";
  }

  private static String getPackageDirs(DataContext dataContext) {
    final Module module = PlatformCoreDataKeys.MODULE.getData(dataContext);
    if (module != null) {
      final VirtualFile[] sourceRoots = ModuleRootManager.getInstance(module).getSourceRoots();
      for (VirtualFile sourceRoot : sourceRoots) {
        // TODO notify if we have multiple source roots and can't build mapping automatically
        final VirtualFile contentRoot = ProjectFileIndex.getInstance(module.getProject()).getContentRootForFile(sourceRoot);
        if (contentRoot != null && !Comparing.equal(contentRoot, sourceRoot)) {
          final String relativePath = VfsUtilCore.getRelativePath(sourceRoot, contentRoot, '/');
          return "\n    package_dir={'': '" + relativePath + "'},";
        }
      }
    }
    return "";
  }

  @Override
  protected PsiDirectory getTargetDirectory(DataContext dataContext, IdeView view) {
    final Module module = PlatformCoreDataKeys.MODULE.getData(dataContext);
    if (module != null) {
      final Collection<VirtualFile> sourceRoots = PyUtil.getSourceRoots(module);
      if (sourceRoots.size() > 0) {
        return PsiManager.getInstance(module.getProject()).findDirectory(sourceRoots.iterator().next());
      }
    }
    return super.getTargetDirectory(dataContext, view);
  }

  @Override
  protected void elementCreated(CreateFromTemplateDialog dialog, PsiElement createdElement) {
    final PropertiesComponent propertiesComponent = PropertiesComponent.getInstance();
    final Properties properties = dialog.getEnteredProperties();
    final String author = properties.getProperty("Author");
    if (author != null) {
      propertiesComponent.setValue(AUTHOR_PROPERTY, author);
    }
    final String authorEmail = properties.getProperty("Author_email");
    if (authorEmail != null) {
      propertiesComponent.setValue(EMAIL_PROPERTY, authorEmail);
    }
  }
}

File: python/src/com/jetbrains/python/packaging/setupPy/CreateSetupPyAction.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.jetbrains.python.packaging.setupPy

import com.intellij.ide.IdeView
import com.intellij.ide.fileTemplates.FileTemplateManager
import com.intellij.ide.fileTemplates.actions.AttributesDefaults
import com.intellij.ide.fileTemplates.actions.CreateFromTemplateAction
import com.intellij.ide.fileTemplates.ui.CreateFromTemplateDialog
import com.intellij.ide.util.PropertiesComponent
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.actionSystem.PlatformCoreDataKeys
import com.intellij.openapi.module.Module
import com.intellij.openapi.roots.ModuleRootManager
import com.intellij.openapi.roots.ProjectFileIndex
import com.intellij.openapi.util.Comparing
import com.intellij.openapi.util.text.StringUtil
import com.intellij.openapi.vfs.VfsUtilCore
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiManager
import com.intellij.util.SystemProperties
import com.jetbrains.python.PyBundle
import com.jetbrains.python.PythonFileType
import com.jetbrains.python.packaging.PyPackageUtil
import com.jetbrains.python.packaging.management.PythonPackageManager
import com.jetbrains.python.packaging.management.hasInstalledPackageSnapshot
import com.jetbrains.python.psi.PyUtil
import com.jetbrains.python.sdk.PythonSdkUtil
import java.util.function.Supplier

class CreateSetupPyAction : CreateFromTemplateAction(
  SETUP_SCRIPT_TEMPLATE_NAME,
  PythonFileType.INSTANCE.icon,
  Supplier { FileTemplateManager.getDefaultInstance().getInternalTemplate(SETUP_SCRIPT_TEMPLATE_NAME) }
) {
  init {
    getTemplatePresentation().setText(PyBundle.message("python.packaging.create.setup.py"))
  }

  override fun update(e: AnActionEvent) {
    val module = e.getData<Module?>(PlatformCoreDataKeys.MODULE)
    e.presentation.setEnabled(module != null && !PyPackageUtil.hasSetupPy(module))
  }

  public override fun getAttributesDefaults(dataContext: DataContext): AttributesDefaults {
    val project = CommonDataKeys.PROJECT.getData(dataContext)
    val defaults = AttributesDefaults("setup.py").withFixedName(true)
    if (project != null) {
      defaults.addPredefined("Import", getSetupImport(dataContext))
      defaults.add("Package_name", project.getName())
      val properties = PropertiesComponent.getInstance()
      defaults.add("Author", properties.getValue(AUTHOR_PROPERTY, SystemProperties.getUserName()))
      defaults.add("Author_email", properties.getValue(EMAIL_PROPERTY, ""))
      defaults.addPredefined("PackageList", getPackageList(dataContext))
      defaults.addPredefined("PackageDirs", getPackageDirs(dataContext))
    }
    defaults.attributeVisibleNames = visibleNames
    return defaults
  }

  override fun getTargetDirectory(dataContext: DataContext, view: IdeView?): PsiDirectory? {
    val module = PlatformCoreDataKeys.MODULE.getData(dataContext)
    if (module != null) {
      val sourceRoots = PyUtil.getSourceRoots(module)
      if (!sourceRoots.isEmpty()) {
        return PsiManager.getInstance(module.getProject()).findDirectory(sourceRoots.iterator().next()!!)
      }
    }
    return super.getTargetDirectory(dataContext, view)
  }

  override fun elementCreated(dialog: CreateFromTemplateDialog, createdElement: PsiElement?) {
    val propertiesComponent = PropertiesComponent.getInstance()
    val properties = dialog.enteredProperties
    val author = properties.getProperty("Author")
    if (author != null) {
      propertiesComponent.setValue(AUTHOR_PROPERTY, author)
    }
    val authorEmail = properties.getProperty("Author_email")
    if (authorEmail != null) {
      propertiesComponent.setValue(EMAIL_PROPERTY, authorEmail)
    }
  }

  companion object {
    private const val AUTHOR_PROPERTY = "python.packaging.author"
    private const val EMAIL_PROPERTY = "python.packaging.author.email"
    const val SETUP_SCRIPT_TEMPLATE_NAME: String = "Setup Script"

    private val visibleNames: MutableMap<String?, String?>
      get() {
        val attributeToName = HashMap<String?, String?>()
        attributeToName.put("Package_name", PyBundle.message("python.packaging.create.setup.package.name"))
        attributeToName.put("Version", PyBundle.message("python.packaging.create.setup.version"))
        attributeToName.put("URL", PyBundle.message("python.packaging.create.setup.url"))
        attributeToName.put("License", PyBundle.message("python.packaging.create.setup.license"))
        attributeToName.put("Author", PyBundle.message("python.packaging.create.setup.author"))
        attributeToName.put("Author_Email", PyBundle.message("python.packaging.create.setup.author.email"))
        attributeToName.put("Description", PyBundle.message("python.packaging.create.setup.description"))
        return attributeToName
      }

    private fun getSetupImport(dataContext: DataContext): String {
      val module = PlatformCoreDataKeys.MODULE.getData(dataContext)
      return if (hasSetuptoolsPackage(module)) "from setuptools import setup" else "from distutils.core import setup"
    }

    private fun hasSetuptoolsPackage(module: Module?): Boolean {
      val sdk = PythonSdkUtil.findPythonSdk(module)
      if (sdk == null) return false

      val project = module?.project ?: return false
      val packageManager = PythonPackageManager.forSdk(project, sdk)

      return packageManager.hasInstalledPackageSnapshot("setuptools")
    }

    private fun getPackageList(dataContext: DataContext): String {
      val module = PlatformCoreDataKeys.MODULE.getData(dataContext)
      if (module != null) {
        return "['" + StringUtil.join(PyPackageUtil.getPackageNames(module), "', '") + "']"
      }
      return "[]"
    }

    private fun getPackageDirs(dataContext: DataContext): String {
      val module = PlatformCoreDataKeys.MODULE.getData(dataContext)
      if (module != null) {
        val sourceRoots = ModuleRootManager.getInstance(module).getSourceRoots()
        for (sourceRoot in sourceRoots) {
          // TODO notify if we have multiple source roots and can't build mapping automatically
          val contentRoot = ProjectFileIndex.getInstance(module.getProject()).getContentRootForFile(sourceRoot)
          if (contentRoot != null && !Comparing.equal<VirtualFile?>(contentRoot, sourceRoot)) {
            val relativePath: String = VfsUtilCore.getRelativePath(sourceRoot, contentRoot, '/')!!
            return "\n    package_dir={'': '" + relativePath + "'},"
          }
        }
      }
      return ""
    }
  }
}